DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c

```diff
@@ -0,0 +1,344 @@
+/** @file
+  TPM2 Table Generator
+
+  Copyright (c) 2022, ARM Limited. All rights reserved.
+  Copyright (c) 2023 - 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+  @par Reference(s):
+  - TCG ACPI 2.0 Specification
+
+  @par Glossary:
+  - Cm or CM   - Configuration Manager
+  - Obj or OBJ - Object
+**/
+
+#include <Library/AcpiLib.h>
+#include <Library/BaseLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/DebugLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Protocol/AcpiTable.h>
+
+// Module specific include files.
+#include <AcpiTableGenerator.h>
+#include <ConfigurationManagerObject.h>
+#include <ConfigurationManagerHelper.h>
+#include <Library/TableHelperLib.h>
+#include <Protocol/ConfigurationManagerProtocol.h>
+#include <IndustryStandard/Tpm2Acpi.h>
+
+#pragma pack(1)
+
+typedef struct {
+  EFI_ACPI_DESCRIPTION_HEADER    Header;
+  UINT16                         PlatformClass;
+  UINT16                         Reserved;
+  UINT64                         AddressOfControlArea;
+  UINT32                         StartMethod;
+  UINT8                          StartMethodParameters[12]; // size up to 12
+  UINT32                         Laml;                      // Optional
+  UINT64                         Lasa;                      // Optional
+} EFI_TPM2_ACPI_TABLE_V4;

```

Comment:
I might be missing information because I didn't find the spec defining the revision 3 of the table, but either:
- This struct should be added to MdePkg/Include/IndustryStandard/Tpm2Acpi.h if it is really different from the revision 3
- Or the StartMethodParameters/Laml/Lasa fields are commented because of the fact that StartMethodParameters is of variable length. In such case you should add the following struct  (to Tpm2Acpi.h) to access them:

typedef struct {
  UINT32                      Laml;                          // Optional
  UINT64                      Lasa;                          // Optional
} SOME_NAME;

Reply:
Actually, this was meant to be a local instantiation of the one defined in Tpm2Acpi.h, so that we could access Laml and Lasa directly. The generic one in Tpm2Acpi.h already has the fields defined for V4 but commented out.

I will spend some time to find a better implementation for this.

Reply:
As it is unknown how long the `StartMethodParameters` field will be in the struct, I don't think this is possible to define a struct to access the Laml/Lasa fields. You will have to typecast a pointer to the end of the `StartMethodParameters` field to the type `SOME_NAME` (cf. above)

Reply:
Actually, according to the spec, "If LAML and LASA are present, then this field is 12 bytes in size." Hence, the offsets to Laml/Lasa are fixed.

Reply:
I see. I will find a better way to implement this.